src/workerd/util/sqlite.c++

```diff
@@ -0,0 +1,835 @@
+// we want to leverage all that code. If we can just make it interpret paths differently, then we
+// can reuse the rest of the implementation.
+
+namespace {
+
+static thread_local int currentVfsRoot = AT_FDCWD;
+// We will tell SQLite to use alternate implementations of path-oriented syscalls which use the
+// `*at()` versions of the calls with `currentVfsRoot` as the directory descriptor. When the
+// descriptor is `AT_FDCWD`, this will naturally reproduce the behavior of the non-`at()` versions.
+// We temporarily swap this for a real desriptor when our custom VFS wrapper is being invoked.
+
+static int replaced_open(const char* path, int flags, int mode) {
+  return openat(currentVfsRoot, path, flags, mode);
+}
+static int replaced_access(const char* path, int type) {
+  return faccessat(currentVfsRoot, path, type, 0);
+}
+static char* replaced_getcwd(char* buf, size_t size) noexcept {
+  KJ_REQUIRE(currentVfsRoot == AT_FDCWD,
+      "SQLite custom VFS shouldn't call getcwd() because we overrode xFullPathname");
+  return getcwd(buf, size);
+}
+static int replaced_stat(const char* path, struct stat* stats) {
+  return fstatat(currentVfsRoot, path, stats, 0);
+}
+static int replaced_unlink(const char* path) {
+  return unlinkat(currentVfsRoot, path, 0);
+}
+static int replaced_mkdir(const char* path, mode_t mode) {
+  return mkdirat(currentVfsRoot, path, mode);
+}
+static int replaced_rmdir(const char* path) {
+  return unlinkat(currentVfsRoot, path, AT_REMOVEDIR);
+}
+static ssize_t replaced_readlink(const char* path, char* buf, size_t len) {
+  return readlinkat(currentVfsRoot, path, buf, len);
+}
+static int replaced_lstat(const char* path, struct stat* stats) {
+  return fstatat(currentVfsRoot, path, stats, AT_SYMLINK_NOFOLLOW);
+}
+
+};
+
+struct Sqlite::Vfs::WrappedNativeFileImpl: public sqlite3_file {
+  // The sqlite3_file implementation we use when wrapping the native filesystem.
+
+  int rootFd;
+
+  // It's expected that the wrapped sqlite_file begins in memory immediately after this object.
+  sqlite3_file* getWrapped() { return reinterpret_cast<sqlite3_file*>(this + 1); }
+
+  static const sqlite3_io_methods METHOD_TABLE;
+};
+
+template <typename Result, typename... Params,
+          Result (*sqlite3_vfs::*slot)(sqlite3_vfs* vfs, Params...)>
+struct Sqlite::Vfs::MethodWrapperHack<Result (*sqlite3_vfs::*)(sqlite3_vfs* vfs, Params...), slot> {
+  // This completely nutso template generates wrapper functions for each of the function pointer
+  // members of sqlite3_vfs. The wrapper function temporarily sets `currentVfsRoot` to the FD
+  // of the directory from the Sqlite::Vfs instance in use, then invokes the same function on
+  // the underlying native VFS.
+
+  static Result wrapper(sqlite3_vfs* vfs, Params... params) noexcept {
+    auto& self = *reinterpret_cast<Sqlite::Vfs*>(vfs->pAppData);
+    KJ_ASSERT(currentVfsRoot == AT_FDCWD);
+    currentVfsRoot = self.rootFd;
+    KJ_DEFER(currentVfsRoot = AT_FDCWD);
```

Comment:
You're correct that exceptions aren't a concern here so it's not really necessary to write in RAII style, but I still prefer it because exception-unsafe code triggers me. (Also it saves the need to store the return value in a temporary variable...)